src/cares_wrap.cc

```diff
@@ -526,6 +551,629 @@ class QueryWrap : public AsyncWrap {
+
+  ares_naptr_reply* current = naptr_start;
+  int offset = ret->Length();
+  for (uint32_t i = 0; current != nullptr; ++i, current = current->next) {
+    Local<Object> naptr_record = Object::New(env->isolate());
+    naptr_record->Set(context,
+                      flags_symbol,
+                      OneByteString(env->isolate(), current->flags)).FromJust();
+    naptr_record->Set(context,
+                      service_symbol,
+                      OneByteString(env->isolate(),
+                                    current->service)).FromJust();
+    naptr_record->Set(context,
+                      regexp_symbol,
+                      OneByteString(env->isolate(),
+                                    current->regexp)).FromJust();
+    naptr_record->Set(context,
+                      replacement_symbol,
+                      OneByteString(env->isolate(),
+                                    current->replacement)).FromJust();
+    naptr_record->Set(context,
+                      order_symbol,
+                      Integer::New(env->isolate(), current->order)).FromJust();
+    naptr_record->Set(context,
+                      preference_symbol,
+                      Integer::New(env->isolate(),
+                                   current->preference)).FromJust();
+    if (need_type)
+      naptr_record->Set(context, type_symbol, naptr_symbol).FromJust();
+
+    ret->Set(context, i + offset, naptr_record).FromJust();
+  }
+
+  ares_free_data(naptr_start);
+  return ARES_SUCCESS;
+}
+
+
+int ParseSoaReply(Environment* env,
+                  unsigned char* buf,
+                  int len,
+                  Local<Object>* ret) {
+  EscapableHandleScope handle_scope(env->isolate());
+  auto context = env->context();
+  Context::Scope context_scope(context);
+
+  /* Can't use ares_parse_soa_reply() here which can only parse single record */
+  unsigned int ancount = cares_get_16bit(buf + 6);
+  unsigned char* ptr = buf + NS_HFIXEDSZ;
+  int rr_type, rr_len;
+  char* name;
+  char* rr_name;
+  long temp_len;  // NOLINT(runtime/int)
+  int status = ares_expand_name(ptr, buf, len, &name, &temp_len);
+  if (status != ARES_SUCCESS) {
+    /* returns EBADRESP in case of invalid input */
+    return status == ARES_EBADNAME ? ARES_EBADRESP : status;
+  }
+
+  if (ptr + temp_len + NS_QFIXEDSZ > buf + len) {
+    free(name);
+    return ARES_EBADRESP;
+  }
+  ptr += temp_len + NS_QFIXEDSZ;
+
+  for (unsigned int i = 0; i < ancount; i++) {
+    status = ares_expand_name(ptr, buf, len, &rr_name, &temp_len);
+
+    if (status != ARES_SUCCESS)
+      break;
+
+    ptr += temp_len;
+    if (ptr + NS_RRFIXEDSZ > buf + len) {
+      free(rr_name);
+      status = ARES_EBADRESP;
+      break;
+    }
+
+    rr_type = cares_get_16bit(ptr);
+    rr_len = cares_get_16bit(ptr + 8);
+    ptr += NS_RRFIXEDSZ;
+
+    /* only need SOA */
+    if (rr_type == ns_t_soa) {
+      ares_soa_reply* soa = node::Malloc<ares_soa_reply>(1);
```

Comment:
The only reason is that this code is referred from C-ARES' SOA parsing function. I'll try to change it to stack-allocated then.